sql/rpl_rli.cc

```diff
@@ -2493,3 +2493,19 @@ void Relay_log_info::adapt_to_master_version(Format_description_log_event *fdle)
     }
   }
 }
+
+void* Relay_log_info::operator new(size_t request)
```

Comment:
@gl-sergei has fixed this in a different way in XB (IIRC the alignment is not required in the first place)

Reply:
He didn't fix this, he removed the alignment requirement from a struct. So that fixes the compilation error but results in a different alignment for these classes, even when allocated on the stack.

My version uses the correct, specified alignment, and I think that's better.

If we need to, I can introduce a helper function instead of these two small functions, which will also be able to handle other platforms - but I checked our supported platform list, and we only support linux.

Reply:
Upstream 8.0 has removed it too, IIRC because the alignment was not shown to have any perf improvement and was always broken (?). If your version is correct, should it be restored in XtraBackup too? Please reach consensus with @gl-sergei.

If this is the correct fix, then check whether this code is not compiled with -fno-exceptions and whether is not supposed to return NULL instead of throwing

Reply:
I didn't check 8.0 before - now I did, and they indeed removed it, but for a different reason: a sanitizer found that they used it without the correct alignment (as mentioned in the commit message: https://github.com/mysql/mysql-server/commit/ebc4a60a46a3077a20ce346dee07bdea2c936bfc ) - and my fix corrects that, after this change, it'll be correctly 64 bit aligned every time.

If it would be compiled without exceptions, that throw would generate an error - and it doesn't, so I think it isn't, but I'll check it to be sure.

Reply:
You are right, this is compiled with exceptions, in which case I am not sure what to do with the caller, which checks for NULL instead of catching the exception.

For 8.0, this will have to be merged with re-enabling the alignment or the fix will be lost

Reply:
Uh, I didn't check that. But that wouldn't work even with the standard new, as that also throws bad_alloc.

I could change it to return NULL, but the standard asks for bad_alloc in this case...

Maybe I could change the factory to catch the exception?

I could also try to submit this patch to mysql, maybe they would accept it if it passes the sanitizer tests?

Reply:
While returning NULL would "work", I don't like introducing broken code to be compatible with other broken code. While we are not a diverging fork and thus any patch has to be evaluated for its maintenance cost (which usually excludes cleanup patches, refactorings, etc), in this particular case catching the exception in the caller seems OK.

Re. upstream submission, a patch to replace NULL-checking with exception catching could be submitted, but it can take up to forever for them to merge it. For 8.0, restoring alignment is a performance patch, they usually want to see some evidence of performance impact